### DISCUSSION TITLE
Adding a custom translation provider in 4.1.6

### DISCUSSION THREAD

**Discussion by gnh1201** [mastodon/mastodon#26870] [Q&A]
I have experience adding a custom translation provider when I was using Mastodon version 4.0.2.

https://github.com/gnh1201/mastodon-lingva

Currently, I am using version 4.1.6.

I tried to add a custom translation provider according to the code for version 4.1.6, but I encountered the following error:

```
Sep  9 20:44:08 gosomi bundle[782892]: [26a0d58f-e22c-4219-9dcd-6a04f953ced0] method=POST path=/api/v1/statuses/111036983556043247/translate format=html controller=Api::V1::Statuses::TranslationsController action=create status=500 error='NoMethodError: undefined method `model_name' for Request:Class
Sep  9 20:44:08 gosomi bundle[782892]: [26a0d58f-e22c-4219-9dcd-6a04f953ced0]
Sep  9 20:44:08 gosomi bundle[782892]: [26a0d58f-e22c-4219-9dcd-6a04f953ced0] NoMethodError (undefined method `model_name' for Request:Class
Sep  9 20:44:08 gosomi bundle[782892]: [26a0d58f-e22c-4219-9dcd-6a04f953ced0]
Sep  9 20:44:08 gosomi bundle[782892]: [26a0d58f-e22c-4219-9dcd-6a04f953ced0] app/controllers/api/v1/statuses/translations_controller.rb:14:in `create'
Sep  9 20:44:08 gosomi bundle[782892]: [26a0d58f-e22c-4219-9dcd-6a04f953ced0] app/controllers/concerns/localized.rb:11:in `set_locale'
Sep  9 20:44:08 gosomi bundle[782892]: [26a0d58f-e22c-4219-9dcd-6a04f953ced0] lib/mastodon/rack_middleware.rb:9:in `call'
```

Despite referencing code from Deepl and LibreTranslate, I'm still encountering this error. What am I missing?

All related files: https://gist.github.com/gnh1201/2ef69f3e57828692e47297edc1c4bb26

**Comment by ryanmio** [comment]
That usually happens when your class is missing that particular method. Since you're customizing the translation provider, make sure your class has the model_name method

**Reply by gnh1201** [reply]
Thanks for your answer. But, neither DeepL nor LibreTranslate classes define a `model_name` method separately. I don't think this changes just because I am adding a new translation provider.

**Reply by ryanmio** [reply]
Ah in that case maybe the issue is in translations_controller.rb. If model_name isn't required in other classes like DeepL or LibreTranslate, then something in your controller logic might be assuming its existence. Let me know if that helps

**Reply by gnh1201** [reply]
Upon reevaluating this issue, I started to consider that I may not have referenced the latest version of the file. (There seems to be a slight discrepancy between the Github Search results and the main branch.) I will review the results obtained after rewriting based on the latest version and then discuss it again.

**Reply by gnh1201** [reply]
Upon checking, I have been confirmed that it follows the style of [version 4.1.6](https://github.com/mastodon/mastodon/blob/v4.1.6/app/lib/translation_service/libre_translate.rb). 

I've checked the `translations_controller.rb` file, and it appears that there is no code dependent on a specific translation provider.

**Comment by ryanmio** [comment]
Have you tried trace that call to model_name mentioned in the error logs? It seems to indicates that something is calling model_name on a request object, but that shouldn't be happening. Apologies I haven't been mor helpful!

**Comment by gnh1201** [comment]
I bypassed the endpoint of the LibreTranslate API included in the mastodon to my server. As a result, I was able to provide my custom translation service without modifying the source code.

However, this does not include the definition of language detection. I think we should change the code that depends on a particular translator vendor to an independent code.
